openssl-sys/build.rs

```diff
@@ -1,86 +1,323 @@
+use std::io::Read;
+use std::path::{Path, PathBuf};
 
 fn main() {
     let target = env::var("TARGET").unwrap();
 
-    // libressl_pnacl_sys links the libs needed.
-    if target.ends_with("nacl") { return; }
+    let openssl_dir = env::var_os("OPENSSL_DIR").unwrap_or_else(|| {
+        find_openssl_dir(&target)
+    });
 
-    let lib_dir = env::var("OPENSSL_LIB_DIR").ok();
-    let include_dir = env::var("OPENSSL_INCLUDE_DIR").ok();
+    let lib_dir = Path::new(&openssl_dir).join("lib");
+    let include_dir = Path::new(&openssl_dir).join("include");
+    if !Path::new(&lib_dir).exists() {
+        panic!("OpenSSL library directory does not exist: {}",
+               lib_dir.to_string_lossy());
+    }
 
-    if lib_dir.is_none() && include_dir.is_none() {
-        // rustc doesn't seem to work with pkg-config's output in mingw64
-        if !target.contains("windows") {
-            if let Ok(info) = pkg_config::find_library("openssl") {
-                // avoid empty include paths as they are not supported by GCC
-                if info.include_paths.len() > 0 {
-                    let paths = env::join_paths(info.include_paths).unwrap();
-                    println!("cargo:include={}", paths.to_str().unwrap());
-                }
-                return;
-            }
+    if !Path::new(&include_dir).exists() {
+        panic!("OpenSSL include directory does not exist: {}",
+               include_dir.to_string_lossy());
+    }
+
+    println!("cargo:rustc-link-search=native={}", lib_dir.to_string_lossy());
+    println!("cargo:include={}", include_dir.to_string_lossy());
+
+    let version = validate_headers(&[include_dir.clone().into()],
+                                   &[lib_dir.clone().into()]);
+
+    let libs = if (version.contains("OpenSSL 1.0.1") ||
+                   version.contains("OpenSSL 1.0.2")) &&
+                  target.contains("windows") {
+        ["ssleay32", "libeay32"]
+    } else if target.contains("windows") {
+        ["libssl", "libcrypto"]
+    } else {
+        ["ssl", "crypto"]
+    };
+
+    let kind = determine_mode(Path::new(&lib_dir), &libs);
+    for lib in libs.iter() {
+        println!("cargo:rustc-link-lib={}={}", kind, lib);
+    }
+}
+
+fn find_openssl_dir(target: &str) -> OsString {
+    let host = env::var("HOST").unwrap();
+
+    if host.contains("apple-darwin") && target.contains("apple-darwin") {
+        let homebrew = Path::new("/usr/local/opt/openssl");
+        if homebrew.exists() {
+            return homebrew.to_path_buf().into()
         }
-        if let Some(mingw_paths) = get_mingw_in_path() {
-            for path in mingw_paths {
-                println!("cargo:rustc-link-search=native={}", path);
-            }
+        let homebrew = Path::new("/usr/local/opt/openssl@1.1");
+        if homebrew.exists() {
+            return homebrew.to_path_buf().into()
+        }
+
+        let system = Path::new("/usr/lib/libssl.0.9.8.dylib");
+        if system.exists() {
```

Comment:
The picture I have in my head is that the OSX codepath behaves just like the Linux codepath except that it first checks for a Homebrew install.